File: packages/embedder/registry.go
package embedder

import (
	"fmt"
	"sort"
	"strings"
)

// Registry maps deployment Profile keys to Embedder implementations (read-only after New).
type Registry struct {
	byProfile map[Profile]Embedder
}

// NewRegistry constructs a registry from profile → impl.
func NewRegistry(byProfile map[Profile]Embedder) (*Registry, error) {
	out := make(map[Profile]Embedder, len(byProfile))
	for profile, emb := range byProfile {
		if err := registerOne(out, profile, emb); err != nil {
			return nil, err
		}
	}
	return &Registry{byProfile: out}, nil
}

func registerOne(dst map[Profile]Embedder, profile Profile, emb Embedder) error {
	key := Profile(strings.TrimSpace(string(profile)))
	if key == "" || !ValidProfile(key) {
		return fmt.Errorf("%w: %q", ErrUnknownProfile, profile)
	}
	if emb == nil {
		return fmt.Errorf("%w: nil embedder for %q", ErrMissingEmbedder, key)
	}
	if got := emb.Profile(); got != key {
		return fmt.Errorf("%w: key %q embedder reports %q", ErrUnknownProfile, key, got)
	}
	if _, dup := dst[key]; dup {
		return fmt.Errorf("%w: %q", ErrDuplicateProfile, key)
	}
	dst[key] = emb
	return nil
}

// ForProfile returns the embedder for profile or ErrUnknownProfile.
func (r *Registry) ForProfile(profile Profile) (Embedder, error) {
	if r == nil {
		return nil, ErrUnknownProfile
	}
	key := Profile(strings.TrimSpace(string(profile)))
	emb, ok := r.byProfile[key]
	if !ok {
		return nil, fmt.Errorf("%w: %q", ErrUnknownProfile, key)
	}
	return emb, nil
}

// Profiles returns sorted profile keys present in the registry.
func (r *Registry) Profiles() []Profile {
	if r == nil {
		return nil
	}
	out := make([]Profile, 0, len(r.byProfile))
	for p := range r.byProfile {
		out = append(out, p)
	}
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}

File: services/embedder/app/registry.py
"""Profile-keyed backend registry (read-only after construction)."""

from __future__ import annotations

from app.backends.base import EmbeddingBackend
from app.errors import (
    DuplicateProfileError,
    MissingBackendError,
    UnknownProfileError,
)
from app.profiles import valid_profile


def _validate_profile_key(profile: str) -> str:
    key = profile.strip()
    if not key or not valid_profile(key):
        raise UnknownProfileError(f"unknown embedding profile: {profile!r}")
    return key


def _validate_backend_present(key: str, backend: EmbeddingBackend | None) -> EmbeddingBackend:
    if backend is None:
        raise MissingBackendError(f"nil backend for {key!r}")
    return backend


def _validate_backend_profile_match(key: str, backend: EmbeddingBackend) -> None:
    if backend.profile != key:
        raise UnknownProfileError(
            f"key {key!r} backend reports {backend.profile!r}"
        )


def _ensure_unique_profile(dst: dict[str, EmbeddingBackend], key: str) -> None:
    if key in dst:
        raise DuplicateProfileError(f"duplicate embedding profile: {key!r}")


class BackendRegistry:
    """Maps deployment profiles to EmbeddingBackend implementations."""

    def __init__(self, by_profile: dict[str, EmbeddingBackend]) -> None:
        registered: dict[str, EmbeddingBackend] = {}
        for profile, backend in by_profile.items():
            self._register_one(registered, profile, backend)
        self._by_profile = registered

    @staticmethod
    def _register_one(
        dst: dict[str, EmbeddingBackend],
        profile: str,
        backend: EmbeddingBackend | None,
    ) -> None:
        key = _validate_profile_key(profile)
        impl = _validate_backend_present(key, backend)
        _validate_backend_profile_match(key, impl)
        _ensure_unique_profile(dst, key)
        dst[key] = impl

    def for_profile(self, profile: str) -> EmbeddingBackend:
        key = profile.strip()
        try:
            return self._by_profile[key]
        except KeyError as exc:
            raise UnknownProfileError(f"unknown embedding profile: {key!r}") from exc

    def profiles(self) -> list[str]:
        return sorted(self._by_profile.keys())
